File: java/srcptest/java/io/github/configur8/ConfigurationTest.java
package io.github.configur8;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import static io.github.configur8.Configuration.ConfigurationTemplate.configurationTemplate;
import static io.github.configur8.Property.string;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

public class ConfigurationTest {

    @Rule
    public ExpectedException thrown = ExpectedException.none();

    @Test
    public void
    returnsDefaultValueOfKeyIfNoneInTheEnvironment() {
        Configuration configuration = configurationTemplate().withProp(string("NOT_IN_ENV"), "The value").reify();
        assertThat(configuration.settings(), hasEntry("NOT_IN_ENV", "The value"));
        assertThat(configuration.valueOf(string("NOT_IN_ENV")), equalTo("The value"));
    }

    @Test
    public void
    prioritisesEnvironmentValueOverDefault() {
        Configuration configuration = configurationTemplate().withProp(string("HOME"), "a default").reify();
        assertThat(configuration.settings(), hasEntry(equalTo("HOME"), not(equalTo("a default"))));
        assertThat(configuration.valueOf(string("HOME")), not(equalTo("a default")));
    }

    @Test
    public void
    throwsIfNoEnvironmentValueOrDefaultForKey() {
        Configuration configuration = configurationTemplate().reify();
        assertThat(configuration.settings(), not(hasKey("MISSING")));

        thrown.expectMessage(containsString("MISSING"));
        thrown.expect(Misconfiguration.class);

        configuration.valueOf(string("MISSING"));
    }

    @Test
    public void
    throwsIfRequiredOverrideIsNotSuppliedWhenCallingReify() {
        Property<String> propertyThatNeedsOverriding = string("REQUIRED_OVERRRIDE");

        thrown.expect(Misconfiguration.class);

        configurationTemplate().requiring(propertyThatNeedsOverriding).reify();
    }
}

File: kotlin/src/test/kotlin/io/github/konfigur8/ConfigurationTest.kt
package io.github.konfigur8

import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import io.github.konfigur8.Property.Companion.int
import io.github.konfigur8.Property.Companion.string
import org.junit.Test
import kotlin.test.fail

class ConfigurationTest {

    private val userProperty = string("bob")
    private val envProperty = string("USER")

    @Test
    @Throws(Exception::class)
    fun blowUpWhenAttemptingToGetAnUnknownProperty() {
        try {
            ConfigurationTemplate().reify().valueOf(string("MISSING"))
            fail("expected exception")
        } catch(e: Misconfiguration) {
            assertThat(e.message!!, equalTo("Unknown configuration key 'MISSING'"))
        }
    }

    @Test
    @Throws(Exception::class)
    fun usesDefaultValueIfNoOverride() {
        val configuration = ConfigurationTemplate().withProp(userProperty, "bill").reify()
        assertThat(configuration.valueOf(userProperty), equalTo("bill"))
        assertThat(configuration[userProperty], equalTo("bill"))
    }

    @Test
    @Throws(Exception::class)
    fun usesEnvironmentValueInPreferenceToDefault() {
        val configuration = ConfigurationTemplate().requiring(envProperty).reify()
        assertThat(configuration.valueOf(envProperty), equalTo(System.getenv(envProperty.name)))
        assertThat(configuration[envProperty], equalTo(System.getenv(envProperty.name)))
    }

    @Test
    @Throws(Exception::class)
    fun canSerialiseAndDeserialiseDefaultValue() {
        val doubler = Property("doubler", { it.toInt() }, { (it * 2).toString() })
        val configuration = ConfigurationTemplate().withProp(doubler, 100).reify()
        assertThat(configuration.valueOf(doubler), equalTo(200))
        assertThat(configuration[doubler], equalTo(200))
    }

    @Test
    @Throws(Exception::class)
    fun usesSystemPropertyValueInPreferenceToAEnvironmentValue() {
        System.setProperty(envProperty.name, "NOTTHEENVUSER")
        val configuration = ConfigurationTemplate().requiring(envProperty).reify()
        assertThat(configuration.valueOf(envProperty), equalTo("NOTTHEENVUSER"))
        assertThat(configuration[envProperty], equalTo("NOTTHEENVUSER"))
    }

    @Test
    fun throwsIfNoValueIsSuppliedAtAll() {
        try {
            ConfigurationTemplate().requiring(userProperty).reify()
            fail("expected exception")
        } catch(e: Misconfiguration) {
            assertThat(e.message!!, equalTo("No value supplied for key 'bob'"))
        }
    }

    @Test
    @Throws(Exception::class)
    fun exposesMapOfPropertiesUsingExposeMode() {
        val privateProperty = string("hello", ExposeMode.Private)
        val settings = ConfigurationTemplate().withProp(privateProperty, "VALUE").reify().settings()
        assertThat(settings, equalTo(mapOf(privateProperty.name to "***********")))
    }

    @Test
    fun iteratesOverStringProperties() {
        val privateProperty = string("private password", ExposeMode.Private)
        val intProperty = int("number of things")
        val properties = ConfigurationTemplate().withProp(privateProperty, "shhhh").withProp(intProperty,123).reify()

        val list: List<Property<*>> = properties.toList()
        val expected: List<Property<*>> = listOf(privateProperty, intProperty)
        assertThat(list,equalTo(expected))
    }

}
